calicoctl/commands/node/status.go

```diff
@@ -129,21 +132,41 @@ func printBGPPeers(ipv string) {
 		os.Exit(1)
 	}
 
-	buf := make([]byte, 1024)
+	// BIRD socket read timeout.
+	timeOut := 2 * time.Second
+	bufReader := bufio.NewReader(c)
 
-	n, err := c.Read(buf[:])
-	if err != nil {
-		fmt.Printf("Error: %v\n", err)
+	birdOut := []string{}
+
+	for {
+		// Set a time-out for reading from the socket connection.
+		// Read operation will fail if no data is received until the timeout.
+		c.SetReadDeadline(time.Now().Add(timeOut))
+
+		// Read string with \n as delim.
+		str, err := bufReader.ReadString('\n')
+		if err != nil {
+			fmt.Println(err)
+			break
+		}
+
+		log.Debugf("Reading line from BIRD sock: %s\n", str)
+
+		// "0000 \n" output from BIRD means end of output.
+		if str == "0000 \n" {
```

Comment:
Better to have the following sort of logic:

-  Read line
-  Line starts with 4-digit code:
   -  Is it the heading code?  Yes, strip off and store the headings (or "assert" that they are what we expect)
   - Is it the end marker?  Yes, exit the loop
   - Is it the start of the data?  Yes?  "assert" we have read the heading.   Strip off code and store the data
  - It's an unexpected code.... Error
-  Line starts with a "space"
   -  Have we started collecting data?  Yes, this is more data, strip the line and store it.
   - We haven't started collecting data, this is unexpected.   Error.


So now we have stored all the data we expect, with no odd 4-digit codes etc.